### PR TITLE
Fixed #29538 -- Added support for ordering related objects by query expressions.

### PR DESCRIPTION
**Pull request by alexandrelaplante**
Added support for ordering related objects by query expressions

https://code.djangoproject.com/ticket/29538

### CODE REVIEW

File: tests/order_with_expression/tests.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        ordering = [models.F('instrument').asc(nulls_first=True)]
+
+        with mock.patch.object(Musician._meta, 'ordering', ordering):
+            self.assertQuerysetEqual(
+                Song.objects.all(),
+                ['Song 3', 'Song 1', 'Song 2'],
+                attrgetter('name')
+            )
+
+    @skip('What should the behavior be?')
```
Comment (alexandrelaplante) [inline tests/order_with_expression/tests.py]: Not sure what the expected behavior here is.

```python
class Album:
    class Meta:
        # Should both of these work the same way? Or does
        # The first mean "order by artist_id"
        # And the second mean, order by the default order of Musician?
        ordering = [models.F('artist').asc(nulls_last=True)]
        ordering = ['artist']
```

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -676,8 +693,14 @@ def find_ordering_name(self, name, opts, alias=None, default_order='ASC',
 
             results = []
             for item in opts.ordering:
-                results.extend(self.find_ordering_name(item, opts, alias,
-                                                       order, already_seen))
+
+                def _join_strings(*strings):
+                    return LOOKUP_SEP.join(filter(None, strings))
```
Comment (orf) [inline django/db/models/sql/compiler.py]: What is this for? Is the `filter()` necessary? Couldn't this just be inlined below?

Comment (alexandrelaplante) [reply django/db/models/sql/compiler.py]: I think this is the alternative
```python
if prefix is None:
    new_prefix = name
else:
    new_prefix = prefix + LOOKUP_SEP + name
```
prefix is `None` at the top of the recursive call.

Which is better?

File: tests/order_with_expression/tests.py (hunk 1 above)
Comment (carltongibson) [reply tests/order_with_expression/tests.py]: `ordering = ['artist']` will mean order by artist (rather than the default order of the related model)

Comment (charettes) [reply tests/order_with_expression/tests.py]: @carltongibson no, you should use`artist_id` if you want to achieve this. [There's a documented difference between `order_by('artist')` and `order_by('artist_id')`](https://docs.djangoproject.com/en/2.1/ref/models/querysets/#order-by) as it was not even possible to disable related ordering before 24ec9538b7ca400f68ba08fab380445ca95d7c02.

Given `Artist.Meta.ordering = ['foo', F('bar').asc(nulls_last=False)]`

then `Album.Meta.ordering = [models.F('artist').asc(nulls_last=True)]`should be the equivalent of

`[F('artist__foo').asc(nulls_last=True), F('artist__bar').asc(nulls_last=True)]`

Comment (carltongibson) [reply tests/order_with_expression/tests.py]: Good. Thanks Simon.

**Comment by alexandrelaplante** [comment]
@carltongibson I tried reusing the models at first but they weren't quite structured in the right way for my tests. I'll give it another try, possibly restructuring the existing tests to accommodate.

**Comment by felixxm** [comment]
Closing due to inactivity.

**Pull request closed without merging**